### ISSUE TITLE
[Adaptive] Setting and managing recognizers for adaptive dialogs

### ISSUE THREAD

**Issue by vishwacsena** [microsoft/botbuilder-dotnet#4651]
Today, we support a variety of recognizer types that work with adaptive dialog. Composer provides UX for a limited set of recognizer types but there is a need for us to have a scalable design to onboard custom/ new recognizer types in both composer as well as in adaptive dialogs.

**Requirements**: 
- Ability for a recognizer type to register to handle technology specific workflows as part of build/ publishing steps in composer.
- Ability to render a scalable UX in composer for the recognizer tied to an adaptive dialog including form UX to edit the properties for a recognizer selection.
- Simplify and streamline the myriad of possible recognizer configurations that are possible today. 
 
**Proposal:**
1. A dialog will always have **only one top level recognizer**. It can either be a `RecognizerSet` or one of the available recognizer types. 
2. In composer, for build/ publishing step, providers can register call backs with their `$kind`. 

**Implications**:
- Proposal #1 means that we will author a new `LuisQnACrossTrainedRecognizer` that holds LUIS and QnA Maker recognizer configuration as properties on it. 
```json
	{
		"recognizer": {
			"$kind": "Microsoft.LuisQnACrossTrainedRecognizer",
			"qnaRecognizer": "dialogName.qna",
			"luisRecognizer": "dialogName.lu"
		}
	}
```
- You can never pick and choose separate recognizer for LU/ QnA. you are always taking both together. If a custom recognizer provides for only one of the functionality (e.g. Orchestrator with intent classification only) then the provider of that recognizer must register a recognizer type that works for both LU and QnA content and choose appropriate defaults/ customization for specific functionality that is not supported within that recognizer type. 
- In composer, we will move the current LUIS, QnA Maker functionality into respective build/ publish call backs tied to `Microsoft.LUISAdaptiveRecognizer` and `Microsoft.QnAMakerRecognizer` etc.
    - For build/ publishing the library could get called with both 'raw' content as well as 'cross-trained' content for that dialog. 
- The recognizer configuration provided by the build step is inlined in the .dialog file. prior to the build step, just the $kind is serialized inline.
    ```json
        {
            "recognizer": {
                "$kind": "Microsoft.LuisQnACrossTrainedRecognizer"
            }
        }
    ````
- The ability to provide multi-lingual support and management need to be exposed as properties by the top level recognizer. E.g. `LuisQnACrossTrainedRecognizer` should internally use and abstract away the `MultiLanguageRecognizer` implementation. 
- The proposal also means composer will expose the following top level recognizers (apart from any custom recognizers added by the user)
    - Regex Recognizer
    - Default Recognizer (LuisQnACrossTrainedRecognizer)
    - Any custom recognizer registered by the user.

**New work (SDK)**:
- `LuisQnACrossTrainedRecognizer`
- `OrchestratorCrossTrainedRecoginzer` 
- [Open] Do we need a multi-lingual RegexRecognizer and switch composer to using that instead?

@tomlm @chrimc62 @cwhitten @boydc2014 @taicchoumsft

**Comment by boydc2014** [comment]
@luhan2017 please also help take a look.

**Comment by luhan2017** [comment]
Just my 2 cents. The runtime don't need to know what is exact recognizer type under CrossTrainedRecognizerSet, it can be qna luis, it can also be ochestrator and pva. Only publish logic need to be aware of that and trigger the corresponding pipeline. 

Can we just add a field for publishing logic? For example，
···
{
		"recognizer": {
			"$kind": "Microsoft.CrossTrainedRecognizer",
                        "builderType": {
                                "isCrossTrainEnabled": true/false,
                                "qna": "qnabuild",
                                "lu": "lubuild/ochestratorbuild/pvabuild"
                         },
			 "recognizers": [
                               "dialogName.qna",
                               "dialogName.lu"
                         ]
		}
	}
···

**Comment by boydc2014** [comment]
> Just my 2 cents. The runtime don't need to know what is exact recognizer type under CrossTrainedRecognizerSet, it can be qna luis, it can also be ochestrator and pva. Only publish logic need to be aware of that and trigger the corresponding pipeline.
> 
> Can we just add a field for publishing logic? For example，
> 
> ```
> {
> 		"recognizer": {
> 			"$kind": "Microsoft.CrossTrainedRecognizer",
>                         "builderType": {
>                                 "isCrossTrainEnabled": true/false,
>                                 "qna": "qnabuild",
>                                 "lu": "lubuild/ochestratorbuild/pvabuild"
>                          },
> 			 "recognizers": [
>                                "dialogName.qna",
>                                "dialogName.lu"
>                          ]
> 		}
> 	}
> ```

This proposal is to not use "CrossTrainedRecoginzerSet" as top-level recognizer, instead, use a LuisQnaCrossTrainedRecognizer. So in orchestrator, it should be a  OrchestratorQnaRecongizer, so that the "$kind" will be able to identify the choice.

**Comment by boydc2014** [comment]
This one is a significant change and requires a design review before move to implementation.

**Comment by boydc2014** [comment]
Add some context about where the original issue, this is discovered when we are designing for the orchestrator integration flow in Composer, in where we found, orchestrator share the same input (.lu file) and deliver the same serialization result (recognizer: dialog.lu). For example 

With both LUIS and orchestrator, you will have 
```
{
 "recognizer": "dialogname.lu"
}
```

then that leave us nowhere we know how to `build` this bot, we need the choice of orchestrator or LUIS serialized somewhere in the dialog in this case. And more generally, the issue is how do we support different type of LU technique and deliver a consistent experience across SDK and Composer.

**Comment by chrimc62** [comment]
I would suggest.
1) We are explicit in the recognizer structure we create, i.e. an explicit recognizer $kind everywhere we need  recognizer.
2) We add an optional source attribute to each recognizer that points to the .lu file it consumes.
3) Build process is associated with the $kind.
This would mean Composer can deal with any recognizer structure and customizing it.  Today it is very confusing because you have to know of the magical connection from .lu reference to a generated recognizer.

**Comment by mrivera-ms** [comment]
@darrenj this seems like a large work item. Could you please prioritize it? Thanks.

**Comment by gabog** [comment]
Hi @boydc2014, @cwhitten, thinks we may need to take this one, can you prioritize and assign it to the next milestone if we are planning to go ahead with it?

**Comment by boydc2014** [comment]
> Hi @boydc2014, @cwhitten, thinks we may need to take this one, can you prioritize and assign it to the next milestone if we are planning to go ahead with it?

Hi, @gabog this was a previous proposal, and was aiming to solve two issues
1. when choosing Orchestrator as recognizer at that time, the serialization result can't differ from LUIS. (because we generated the recognizer dialog at very late when building bots at that time)
2. to make composer's authored bots be CI/CD friendly. 

We organized a review meeting with ChrisMc, Tom, ChrisW. Eventually we choose another way to solve those issues.  Specially,  without changing the top-level recognizer type or hierarchy, we choose to serialize out all the **recognizer dialogs** into a "recognizers" folder(shipped in 1.3),  so that, composer can look into any type in the serialized recognizer tree, and get what we need. 

With that said, this proposal can still help in certain cases, but not a blocker for composer anymore, and we leave this as backlog. So, is there any new scenario driving us to pick this up? If so, sure i can take the implementation this for sure.
